File: typescript/transducers.ts
import {
    comp,
    iterator,
    map,
    mapcat,
    mapIndexed,
    partitionWhen,
} from "@thi.ng/transducers";

interface Section {
    title: string;
    reset_lesson_position?: boolean;
    lessons: { name: string; position?: number }[];
    position?: number;
}

/**
 * Source data
 */
const sections: Section[] = [
    {
        title: "Getting started",
        reset_lesson_position: false,
        lessons: [{ name: "Welcome" }, { name: "Installation" }],
    },

    {
        title: "Basic operator",
        reset_lesson_position: false,
        lessons: [
            { name: "Addition / Subtraction" },
            { name: "Multiplication / Division" },
        ],
    },

    {
        title: "Advanced topics",
        reset_lesson_position: true,
        lessons: [{ name: "Mutability" }, { name: "Immutability" }],
    },
];

/**
 * Helper function to immutably set `key` in given `obj`ect.
 *
 * @param obj
 * @param key
 * @param val
 */
const setKey = <T, K extends keyof T>(obj: T, key: keyof T, val: T[K]) => ({
    ...obj,
    [key]: val,
});

const transformed = [
    ...iterator(
        // build a composed multi-stage transformation function (transducer)
        comp(
            // group sections until next reset flag
            partitionWhen((section) => !!section.reset_lesson_position),
            // transform each group of sections w/ its own internal position counter
            // however `mapcat` (unlike `map`) yields a flat sequence of results
            mapcat((group) => {
                // lesson-within-group counter
                let offset = 1;
                return map((section) => {
                    const pos = offset;
                    offset += section.lessons.length;
                    // immutably label lessons
                    return setKey(
                        section,
                        "lessons",
                        section.lessons.map((l, i) =>
                            setKey(l, "position", pos + i)
                        )
                    );
                }, group);
            }),
            // label/number sections themselves (start from #1)
            mapIndexed((i, section) => setKey(section, "position", i), 1)
        ),
        sections
    ),
];

console.log(JSON.stringify(transformed, null, 4));

File: clojure/transducers.clj
(ns clojure.transducers)

(def data [{:title "Getting started"
            :reset_lesson_position false
            :lessons [{:name "Welcome"}
                      {:name "Installation"}]}
           {:title "Basic operator"
            :reset_lesson_position false
            :lessons [{:name "Addition / Subtraction"}
                      {:name "Multiplication / Division"}]}
           {:title "Advanced topics"
            :reset_lesson_position true
            :lessons [{:name "Mutability"}
                      {:name "Immutability"}]}])

(defn split-each [pred]
  (comp
   (mapcat (fn [section]
             (if (pred section)
               [:restart section]
               [section])))
   (partition-by #{:restart})
   (remove #{[:restart]})))

(def number-sections&lessons
  (comp
   (map-indexed (fn [idx section]
                  (assoc section :position (inc idx))))
   (split-each :reset_lesson_position)
   (mapcat (fn [sections]
             (eduction
              (mapcat (fn [section]
                        (eduction
                         (map (fn [lesson]
                                (assoc lesson :section section)))
                         (get section :lessons))))
              (map-indexed (fn [idx lesson]
                             (assoc lesson :position (inc idx))))
              (partition-by :section)
              (map (fn [lessons-with-section]
                     (let [section (:section (first lessons-with-section))
                           lessons (mapv #(dissoc % :section) lessons-with-section)]
                       (assoc section :lessons lessons))))
              sections)))))

(def sections
  (into []
        number-sections&lessons
        data))

(println sections)


(comment
  (def clj-out
    [{:title "Getting started"
      :reset_lesson_position false,
      :lessons [{:name "Welcome" :position 1}
                {:name "Installation" :position 2}],
      :position 1}
     {:title "Basic operator"
      :reset_lesson_position false,
      :lessons [{:name "Addition / Subtraction" :position 3}
                {:name "Multiplication / Division" :position 4}],
      :position 2}
     {:title "Advanced topics"
      :reset_lesson_position true,
      :lessons [{:name "Mutability" :position 1}
                {:name "Immutability" :position 2}],
      :position 3}])

  (= sections clj-out) ;;=> true

  )
